src/Roslyn.Diagnostics.Analyzers/UnitTests/CSharpDoNotCapturePrimaryConstructorParametersTests.cs

```diff
@@ -0,0 +1,213 @@
+
+            await new VerifyCS.Test
+            {
+                TestCode = source,
+                LanguageVersion = Microsoft.CodeAnalysis.CSharp.LanguageVersion.CSharp12,
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task ErrorOnCapture_InIndexer()
+        {
+            var source = """
+                class C(int i)
+                {
+                    private int this[int param]
+                    {
+                        get => [|i|];
+                        set => [|i|] = value;
+                    }
+                }
+                """;
+
+            await new VerifyCS.Test
+            {
+                TestCode = source,
+                LanguageVersion = Microsoft.CodeAnalysis.CSharp.LanguageVersion.CSharp12,
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task ErrorOnCapture_InEvent()
+        {
+            var source = """
+                class C(int i)
+                {
+                    public event System.Action E
+                    {
+                        add => _ = [|i|];
+                        remove => _ = [|i|];
+                    }
+                }
+                """;
+
+            await new VerifyCS.Test
+            {
+                TestCode = source,
+                LanguageVersion = Microsoft.CodeAnalysis.CSharp.LanguageVersion.CSharp12,
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task ErrorOnCapture_UseInSubsequentConstructor()
+        {
+            var source = """
+                class C(int i)
+                {
+                    C(bool b) : this(1)
+                    {
+                        _ = i;
+                    }
+                }
+                """;
+
+            await new VerifyCS.Test
+            {
+                TestCode = source,
+                LanguageVersion = Microsoft.CodeAnalysis.CSharp.LanguageVersion.CSharp12,
+                ExpectedDiagnostics = {
+                    // /0/Test0.cs(5,13): error RS0103: Primary constructor parameter 'i' should not be implicitly captured
+                    VerifyCS.Diagnostic().WithSpan(5, 13, 5, 14).WithArguments("i"),
+                    // /0/Test0.cs(5,13): error CS9105: Cannot use primary constructor parameter 'int i' in this context.
+                    DiagnosticResult.CompilerError("CS9105").WithSpan(5, 13, 5, 14).WithArguments("int i"),
+                }
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task NoError_PassToBase()
+        {
+            var source = """
+                class Base(int i);
+                class Derived(int i) : Base(i);
+                """;
+
+            await new VerifyCS.Test
+            {
+                TestCode = source,
+                LanguageVersion = Microsoft.CodeAnalysis.CSharp.LanguageVersion.CSharp12,
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task NoError_FieldInitializer()
```

Comment:
>NoError_FieldInitializer [](http://example.com/codeflow?start=26&length=24)

I must be missing something obvious. It looks like the only difference between this unit-test and ```ErrorOnCapture_InMethod``` is the test scenario. This test is supposed to assert no error and ```ErrorOnCapture_InMethod``` is supposed to assert an error. How does this work?

Reply:
Assuming what you're asking about is how `ErrorOnCapture_InMethod` is asserting that there's an error, the `[| |]` spans mark where the expected analyzer diagnostic is supposed to happen. The testing framework preprocesses the source, removing those markers and then asserting that all spans have matching diagnostics, and that there are no extra diagnostics left over after accounting for all expected diagnostics.